src/cbmc/bmc.cpp

```diff
@@ -355,72 +355,12 @@ void bmct::setup()
     setup_unwind();
 }
 
-safety_checkert::resultt bmct::execute(const goto_functionst &goto_functions)
+safety_checkert::resultt bmct::run_and_catch_exceptions(
+  std::function<safety_checkert::resultt(void)> torun)
 {
   try
   {
-    // perform symbolic execution
-    symex(goto_functions);
-
-    // add a partial ordering, if required
-    if(equation.has_threads())
-    {
-      memory_model->set_message_handler(get_message_handler());
-      (*memory_model)(equation);
-    }
-
-  statistics() << "size of program expression: "
-               << equation.SSA_steps.size()
-               << " steps" << eom;
-
-    slice();
-
-    // coverage report
-    std::string cov_out=options.get_option("symex-coverage-report");
-    if(!cov_out.empty() &&
-       symex.output_coverage_report(goto_functions, cov_out))
-    {
-      error() << "Failed to write symex coverage report" << eom;
-      return safety_checkert::resultt::ERROR;
-    }
-
-    if(options.get_bool_option("show-vcc"))
-    {
-      show_vcc();
-      return safety_checkert::resultt::SAFE; // to indicate non-error
-    }
-
-    if(!options.get_list_option("cover").empty())
-    {
-      const optionst::value_listt criteria=
-        options.get_list_option("cover");
-      return cover(goto_functions, criteria)?
-        safety_checkert::resultt::ERROR:safety_checkert::resultt::SAFE;
-    }
-
-    if(options.get_option("localize-faults")!="")
-    {
-      fault_localizationt fault_localization(
-        goto_functions, *this, options);
-      return fault_localization();
-    }
-
-    // any properties to check at all?
-    if(!options.get_bool_option("program-only") &&
-       symex.remaining_vccs==0)
-    {
-      report_success();
-      output_graphml(resultt::SAFE, goto_functions);
-      return safety_checkert::resultt::SAFE;
-    }
-
-    if(options.get_bool_option("program-only"))
-    {
-      show_program();
-      return safety_checkert::resultt::SAFE;
-    }
-
-    return decide(goto_functions, prop_conv);
+    return torun();
```

Comment:
Why did you need to rewrite the entire function when seemingly just the type of the parameter should change?

Reply:
I needed to provide two alternate versions of `execute` (see below this function), which have different prefixes depending on whether they're using a lazy or an eager goto model, but then converge after symex. This particular factoring was to avoid duplicating the error-handling code between the two cases.